### PR TITLE
Python 2to3, too

### PR DESCRIPTION
**Pull request by grahamc**
Continues #1214 by @taneb. The result of running 2to3 on the codebase, plus manual fixes when we're dealing with bytes and strings... plus a few minor 2-to-3 incompatibilities like Queue.Queue and output flushing.

I have tested this deploying to machines which are deployed without any plugin, as well as machines deployed with the nixops-packet plugin, using:

* this branch of nixops-packet https://github.com/input-output-hk/nixops-packet/compare/master...grahamc:py3 (only the last 3 commits are relevant)
*  a recent nixpkgs-unstable at b94c1c89f69563a9fc2ceee487b9bc19e5234d6a
* and this Nixpkgs overlay:

```nix
    (final: super: {
      nixopsBootstrap = (import "${../../../github.com/NixOS/nixops}/release.nix" {
        nixpkgs = final.path;
        p = (nixopsPlugins: [
        ]);
      }).build."${system}";

      nixops-packet = (import "${../../../github.com/input-output-hk/nixops-packet}/release.nix" {
        nixopsSrc = {
          outPath = ../../../github.com/NixOS/nixops;
          revCount = 0;
          shortRev = "abcdef";
          rev = "HEAD";
        };
        nixpkgs = final.path;
        nixops = final.nixopsBootstrap;
      });

      nixops = (import "${../../../github.com/NixOS/nixops}/release.nix" {
        nixpkgs = final.path;
        p = (nixopsPlugins: [
          final.nixops-packet
        ]);
      }).build."${system}";
    })
```

I have tested these things:

* Creating a new system
* Destroying an unneeded system
* `nixops ssh-for-each`
* `nixops deploy`
* `nixops deploy --force-reboot --exclude foo bar`
* `nixops deploy -k`
* `nixops deploy --include foo bar`
* `nixops info`
* `nixops list`
* `nixops check`
* `nixops send-keys`
* `nixops show-arguments` (but I had none)
* `nixops show-physical`
* `nixops ssh` (both to and from)
* `nixops dump-nix-paths`
* `nixops export`

I fixed many problems with combining bytes and strings in the logger, and bytes and strings when running remote commands. I fixed the stdout handler which had trouble with `b"" != ""` especially when detecting a remote SSH command was completed writing. I fixed the logger not flushing after each `.` write, and not flushing when asking if destroying things is okay. 

IMO this is ready to go, and the upgrades to other plugins will need to be done after this merges.

### CODE REVIEW

**Comment by AmineChikhaoui** [comment]
had to modify `script/nixops` to have python3 in the shebang.
And got this error later:
````
  File "/home/amine/src/nixops/nixops/deployment.py", line 741, in worker
    res = m.switch_to_configuration(switch_method, sync)
  File "/home/amine/src/nixops/nixops/backends/none.py", line 65, in switch_to_configuration
    res = super(NoneState, self).switch_to_configuration(method, sync, command)
  File "/home/amine/src/nixops/nixops/backends/__init__.py", line 351, in switch_to_configuration
    return self.run_command(cmd, check=False)
  File "/home/amine/src/nixops/nixops/backends/__init__.py", line 337, in run_command
    return self.ssh.run_command(command, self.get_ssh_flags(), **kwargs)
  File "/home/amine/src/nixops/nixops/ssh_util.py", line 290, in run_command
    return nixops.util.logged_exec(cmd, self._logger, **kwargs)
  File "/home/amine/src/nixops/nixops/util.py", line 136, in logged_exec
    logger.log_start(data[start:])
  File "/home/amine/src/nixops/nixops/logger.py", line 145, in log_start
    self.main_logger.log_start(self._log_prefix, msg)
  File "/home/amine/src/nixops/nixops/logger.py", line 40, in log_start
    self._log_file.write(msg)
  File "/home/amine/src/nixops/nixops/util.py", line 293, in write
    self.stderr.write(data)
TypeError: write() argument must be str, not bytes
```

**Comment by grahamc** [comment]
:clown_face: I never ran it like that in dev mode, I used the `nix-build`'t version :grimacing: I'll take another crack.

File: tests/functional/single_machine_test.py [hunk 1]
```diff
@@ -46,5 +46,5 @@ def test_libvirtd(self):
 
     def check_command(self, command):
         self.depl.evaluate()
-        machine = self.depl.machines.values()[0]
+        machine = list(self.depl.machines.values())[0]
```
Comment (multun) [inline tests/functional/single_machine_test.py]: `next(iter(self.depl.machines.values()))`

File: tests/functional/test_backups.py [hunk 2]
```diff
@@ -64,5 +64,5 @@ def backup_and_restore_path(self, path=""):
 
     def check_command(self, command):
         self.depl.evaluate()
-        machine = self.depl.machines.values()[0]
+        machine = list(self.depl.machines.values())[0]
```
Comment (multun) [inline tests/functional/test_backups.py]: `next(iter(self.depl.machines.values()))`

File: tests/functional/test_ec2_with_nvme_device_mapping.py [hunk 3]
```diff
@@ -26,5 +26,5 @@ def test_ec2_with_nvme_device_mapping(self):
 
     def check_command(self, command):
         self.depl.evaluate()
-        machine = self.depl.machines.values()[0]
+        machine = list(self.depl.machines.values())[0]
```
Comment (multun) [inline tests/functional/test_ec2_with_nvme_device_mapping.py]: `next(iter(self.depl.machines.values()))`

File: tests/functional/test_rebooting_reboots.py [hunk 4]
```diff
@@ -9,7 +9,7 @@ def run_check(self):
         self.depl.deploy()
         self.check_command("touch /run/not-rebooted")
         self.depl.reboot_machines(wait=True)
-        m = self.depl.active.values()[0]
+        m = list(self.depl.active.values())[0]
```
Comment (multun) [inline tests/functional/test_rebooting_reboots.py]: `next(iter(self.depl.machines.values()))`

File: tests/functional/test_stopping_stops.py [hunk 5]
```diff
@@ -6,5 +6,5 @@ class TestStoppingStops(single_machine_test.SingleMachineTest):
     def run_check(self):
         self.depl.deploy()
         self.depl.stop_machines()
-        m = self.depl.active.values()[0]
+        m = list(self.depl.active.values())[0]
```
Comment (multun) [inline tests/functional/test_stopping_stops.py]: `next(iter(self.depl.machines.values()))`

File: nixops/deployment.py [hunk 6]
```diff
@@ -97,15 +97,15 @@ def tempdir(self):
 
     @property
     def machines(self):
-        return {n: r for n, r in self.resources.items() if is_machine(r)}
+        return {n: r for n, r in list(self.resources.items()) if is_machine(r)}
```
Comment (multun) [inline nixops/deployment.py]: same

File: nixops/deployment.py [hunk 7]
```diff
@@ -97,15 +97,15 @@ def tempdir(self):
...
 
     @property
     def active(self): # FIXME: rename to "active_machines"
-        return {n: r for n, r in self.resources.items() if is_machine(r) and not r.obsolete}
+        return {n: r for n, r in list(self.resources.items()) if is_machine(r) and not r.obsolete}
```
Comment (multun) [inline nixops/deployment.py]: same

File: nixops/deployment.py [hunk 8]
```diff
@@ -97,15 +97,15 @@ def tempdir(self):
...
 
     @property
     def active_resources(self):
-        return {n: r for n, r in self.resources.items() if not r.obsolete}
+        return {n: r for n, r in list(self.resources.items()) if not r.obsolete}
```
Comment (multun) [inline nixops/deployment.py]: same

File: nixops/deployment.py [hunk 9]
```diff
@@ -529,15 +529,15 @@ def do_machine(m):
...
+        for m in active_machines.values():
             do_machine(m)
 
         def emit_resource(r):
             config = []
             config.extend(attrs_per_resource[r.name])
             if is_machine(r):
                 # Sort the hosts by its canonical host names.
-                sorted_hosts = sorted(hosts[r.name].iteritems(),
+                sorted_hosts = sorted(iter(hosts[r.name].items()),
```
Comment (multun) [inline nixops/deployment.py]: `iter()` isn't needed

File: nixops/deployment.py [hunk 10]
```diff
@@ -789,22 +789,22 @@ def _get_free_resource_index(self):
...
         self.evaluate_active(include, exclude) # unnecessary?
         machine_backups = {}
-        for m in self.active.itervalues():
+        for m in self.active.values():
             if should_do(m, include, exclude):
                 machine_backups[m.name] = m.get_backups()
 
         # merging machine backups into network backups
-        backup_ids = [b for bs in machine_backups.values() for b in bs.keys()]
+        backup_ids = [b for bs in list(machine_backups.values()) for b in list(bs.keys())]
```
Comment (multun) [inline nixops/deployment.py]: same here

File: nixops/deployment.py [hunk 11]
```diff
@@ -789,22 +789,22 @@ def _get_free_resource_index(self):
...
             backups[backup_id] = {}
             backups[backup_id]['machines'] = {}
             backups[backup_id]['info'] = []
             backups[backup_id]['status'] = 'complete'
             backup = backups[backup_id]
-            for m in self.active.itervalues():
+            for m in self.active.values():
                 if should_do(m, include, exclude):
-                    if backup_id in machine_backups[m.name].keys():
+                    if backup_id in list(machine_backups[m.name].keys()):
```
Comment (multun) [inline nixops/deployment.py]: same

File: nixops/deployment.py [hunk 12]
```diff
@@ -818,7 +818,7 @@ def get_backups(self, include=[], exclude=[]):
 
     def clean_backups(self, keep, keep_days, keep_physical = False):
         _backups = self.get_backups()
-        backup_ids = [b for b in _backups.keys()]
+        backup_ids = [b for b in list(_backups.keys())]
```
Comment (multun) [inline nixops/deployment.py]: same

File: nixops/deployment.py [hunk 13]
```diff
@@ -889,7 +889,7 @@ def evaluate_active(self, include=[], exclude=[], kill_obsolete=False):
         # Determine the set of active resources.  (We can't just
         # delete obsolete resources from ‘self.resources’ because they
         # contain important state that we don't want to forget about.)
-        for m in self.resources.values():
+        for m in list(self.resources.values()):
```
Comment (multun) [inline nixops/deployment.py]: same

File: nixops/deployment.py [hunk 14]
```diff
@@ -1083,7 +1083,7 @@ def _rollback(self, generation, include=[], exclude=[], check=False,
             names.add(filename)
 
         # Update the set of active machines.
-        for m in self.machines.values():
+        for m in list(self.machines.values()):
```
Comment (multun) [inline nixops/deployment.py]: same

**Comment by grahamc** [comment]
@multun all of those were added by 2to3. I'm a bit spooked about making those changes -- but trust your expertise. Would you mind sending my branch a PR?

**Comment by multun** [comment]
yeah sure ! I should probably have figured that out instead of making way too many comments, sorry about that

**Comment by grahamc** [comment]
No worries :D I appreciate you taking the time for it. Ping me here when you've opened that PR?

**Comment by grahamc** [comment]
@AmineChikhaoui I had to change the release.nix to create a nixops without any plugins, pass that to the plugins for testing, and then create a nixops combining them all together. Then I updated scripts/nixops to use python3 and verified it works ... can you try again?

**Comment by AmineChikhaoui** [comment]
@grahamc thanks seems to be working now !

**Comment by grahamc** [comment]
I rebased this on to the reformatted master, since reformatting & a mechanical 2->3 operation is probably a good thing to bulk together.

**Comment by dhess** [comment]
I've tested this with 4 of my own deployments (mainly `deploy`) and it's working great for me.

**Comment by grahamc** [comment]
Note, other people can test this PR too. If you only use the `none` backend, you can simply:

```
nix-build ./release.nix -A build.x86_64-linux
```

and use the resulting ./result/bin/nixops.

**Comment by grahamc** [comment]
btw I'm planning on merging this tomorrow.

**Comment by elseym** [comment]
just tested this on darwin, got the following error:
```
building all machine configurations...
Traceback (most recent call last):
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/bin/.nixops-wrapped", line 251, in <module>
    args.op(args)
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/script_defs.py", line 571, in op_deploy
    max_concurrent_activate=args.max_concurrent_activate,
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/deployment.py", line 1347, in deploy
    self.run_with_notify("deploy", lambda: self._deploy(**kwargs))
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/deployment.py", line 1336, in run_with_notify
    f()
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/deployment.py", line 1347, in <lambda>
    self.run_with_notify("deploy", lambda: self._deploy(**kwargs))
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/deployment.py", line 1267, in _deploy
    dry_run=dry_run, repair=repair, include=include, exclude=exclude
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/deployment.py", line 772, in build_configs
    key_file = m.get_ssh_private_key_file()
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/backends/none.py", line 93, in get_ssh_private_key_file
    return self.write_ssh_private_key(self._ssh_private_key)
  File "/nix/store/3jp88jwfpy4nabx8fwh970mpdacihp0l-nixops-1.8pre0_abcdef/lib/python3.7/site-packages/nixops/backends/__init__.py", line 375, in write_ssh_private_key
    f.write(private_key)
TypeError: write() argument must be str, not None
```

we suspect that this is because not all hosts in the deployment are in the state file yet.

/cc @lheckemann

**Comment by lheckemann** [comment]
>we suspect that this is because not all hosts in the deployment are in the state file yet.

we just realised: this is correct but has nothing to do with the python3 changes and happens with `--build-only` in the python2 version if the hosts have never been deployed before as well.

**Pull request closed without merging**